runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/functions/GroupAlsoByWindowViaOutputBufferFn.java

```diff
@@ -65,9 +65,15 @@ public GroupAlsoByWindowViaOutputBufferFn(
 
   @Override
   public Iterator<WindowedValue<KV<K, Iterable<InputT>>>> call(
-      KV<K, Iterable<WindowedValue<InputT>>> kv) throws Exception {
-    K key = kv.getKey();
-    Iterable<WindowedValue<InputT>> values = kv.getValue();
+      K key, Iterator<WindowedValue<KV<K, InputT>>> iterator) throws Exception {
+
+    // we have to meterialize the Iterator because ReduceFnRunner.processElements expects
+    // ArrayList<WindowedValue<InputT>> and not Iterator<WindowedValue<KV<K, InputT>>>
```

Comment:
It expects `Iterable` not `ArrayList`. It looks like doing the refactoring in `ReduceFnRunner#processElements` to support `Iterator` instead could be a nice fix and avoid this extra materialization for the Spark runner (maybe worth to dig deeper there in a different issue/JIRA).

Reply:
I have already taken a look. Each ReduceFnRunner instance works in parallel in the parallel groups of the flatmapGroups and works on a materialization of the elements of the group. See among other things ReduceFnRunner line 326: reduceFnRunner collects all the windows to merge them. Changing the signature of the processElement method will not change the need for materialization. The important is to avoid materializing the whole collection in one place (which we do not do). We materialize only the part of the collection for each group.